pkg/workflow/workflow.go

```diff
@@ -2,25 +2,216 @@ package workflow
+
+	nodes           map[string]v1alpha12.NodeStatus
+	forwardWorkflow *v1alpha12.Workflow
+	reverseWorkflow *v1alpha12.Workflow
+}
+
+func NewBuilder(client client.Client, logger logr.Logger) *Builder {
+	return &Builder{
+		client:  client,
+		logger:  logger,
+		options: ClientOptions{},
+	}
+}
+
+func (builder *Builder) Forward(appGroup *v1alpha1.ApplicationGroup) *Builder {
+	builder.clientType = Forward
+	builder.appGroup = appGroup
+	return builder
+}
+
+func (builder *Builder) Reverse(forwardWorkflow *v1alpha12.Workflow, nodes map[string]v1alpha12.NodeStatus) *Builder {
+	builder.clientType = Reverse
+	builder.forwardWorkflow = forwardWorkflow
+	builder.nodes = nodes
+	return builder
+}
+
+func (builder *Builder) WithParallelism(numNodes int64) *Builder {
+	builder.options.parallelism = &numNodes
+	return builder
+}
+
+func (builder *Builder) WithStagingRepo(stagingURL string) *Builder {
+	builder.options.stagingRepo = stagingURL
+	return builder
+}
+
+func (builder *Builder) InNamespace(namespace string) *Builder {
+	builder.options.namespace = namespace
+	return builder
+}
+
+func (builder *Builder) WithExecutor(executor ExecutorFunc) *Builder {
+	builder.executor = executor
+	return builder
+}
+
+func (builder *Builder) Build() (Client, error) {
+	switch builder.clientType {
+	case Forward:
+		forwardClient := &ForwardWorkflowClient{
+			Client:        builder.client,
+			Logger:        builder.logger,
+			ClientOptions: builder.options,
+			appGroup:      builder.appGroup,
+			executor:      builder.executor,
+		}
+		if builder.executor == nil {
+			forwardClient.executor = defaultExecutor
+		}
+		return forwardClient, nil
+	case Reverse:
+		reverseClient := &ReverseWorkflowClient{
+			Client:          builder.client,
+			Logger:          builder.logger,
+			ClientOptions:   builder.options,
+			forwardWorkflow: builder.forwardWorkflow,
+			nodes:           builder.nodes,
+			executor:        builder.executor,
+		}
+		if builder.executor == nil {
+			reverseClient.executor = defaultExecutor
+		}
+		return reverseClient, nil
+	}
+	return nil, fmt.Errorf("failed to build engine because type wasn't specified")
+}
+
+func Run(ctx context.Context, engine Client) error {
+	if err := engine.Generate(); err != nil {
+		engine.GetLogger().Error(err, "engine failed to generate workflow")
+		return fmt.Errorf("failed to generate workflow : %w", err)
+	}
+	if err := engine.Submit(ctx); err != nil {
+		engine.GetLogger().Error(err, "engine failed to submit reverse workflow")
+		return err
+	}
+	return nil
+}
+
+func Suspend(ctx context.Context, engine Client) error {
+	// suspend a workflow if it is not already finished or suspended
+	workflow, err := engine.GetWorkflow(ctx)
+	if client.IgnoreNotFound(err) != nil {
+		return err
+	} else if err != nil || !workflow.Status.FinishedAt.IsZero() {
+		engine.GetLogger().Info("forward workflow not found, no need to suspend")
+		return nil
+	}
+	if workflow.Spec.Suspend == nil || !*workflow.Spec.Suspend {
```

Comment:
Does suspending a completed workflow have any side-effects?

Reply:
No it shouldn't have side effects since we aren't re-using the workflow object